src/tree/tree_model.cc

```diff
@@ -617,6 +620,82 @@ std::string RegTree::DumpModel(const FeatureMap& fmap,
   return result;
 }
 
+void RegTree::Load(Json const& in) {
+  fromJson(in["tree_param"], &param);
+  {
+    std::vector<Json> const& v_j_stats = get<Array>(in["stats"]);
+    stats_.resize(v_j_stats.size());
+
+    for (size_t i = 0; i < stats_.size(); ++i) {
+      std::vector<Json> const& j_stat = get<Array const>(v_j_stats[i]);
+      auto& s = stats_[i];
+      s.loss_chg = get<Number const>(j_stat[0]);
+      s.sum_hess = get<Number const>(j_stat[1]);
+      s.base_weight = get<Number const>(j_stat[2]);
+      s.leaf_child_cnt = get<Integer const>(j_stat[3]);
+    }
+    CHECK_EQ(param.num_nodes, stats_.size());
+  }
+
+  {
+    std::vector<Json> const& v_j_nodes = get<Array>(in["nodes"]);
+    nodes_.resize(v_j_nodes.size());
+
+    for (size_t i = 0; i < nodes_.size(); ++i) {
+      auto const& j_node = get<Array const>(v_j_nodes[i]);
+      auto& n = nodes_[i];
+      auto left = get<Integer const>(j_node[0]);
+      auto right = get<Integer const>(j_node[1]);
+      auto parent = get<Integer const>(j_node[2]);
+      auto split_ind = get<Integer const>(j_node[3]);
+      auto split_cond = get<Number const>(j_node[4]);
+      auto default_left = get<Boolean const>(j_node[5]);
+      n = Node(left, right, parent, split_ind, split_cond, default_left);
+    }
+    CHECK_EQ(param.num_nodes, nodes_.size());
+  }
+}
+
+void RegTree::Save(Json* p_out) const {
+  auto& out = *p_out;
+  CHECK_EQ(param.num_nodes, static_cast<int>(nodes_.size()));
+  CHECK_EQ(param.num_nodes, static_cast<int>(stats_.size()));
+  out["tree_param"] = toJson(param);
+  CHECK_EQ(get<String>(out["tree_param"]["num_nodes"]), std::to_string(param.num_nodes));
+
+  std::vector<Json> v_j_stats(stats_.size());
+  using I = Integer::Int;
+
+  for (size_t i = 0; i < stats_.size(); ++i) {
+    auto const& s = stats_[i];
+    std::vector<Json> j_stats(4);
+    j_stats[0] = s.loss_chg;
+    j_stats[1] = s.sum_hess;
+    j_stats[2] = s.base_weight;
+    j_stats[3] = static_cast<I>(s.leaf_child_cnt);
+
+    v_j_stats[i] = std::move(j_stats);
+  }
+  out["stats"] = std::move(v_j_stats);
+
+  std::vector<Json> v_j_nodes(nodes_.size());
+  for (size_t i = 0; i < nodes_.size(); ++i) {
```

Comment:
What is a motivation of representing a tree node as an array?

I see the following pros and cons.

Pros.
1. Less space
2. Faster to parse
3. Easier to parse: no need to deal with missing attributes

Cons
1. It is harder to evolve. What if we want to add another attribute? Just add it to the end? What is we want to deprecate the element? Should we still put a dummy value?
2. The code is much harder to read. We need to remember to which index represent which attribute

So in the end we can evolve the schema even with this being an array (much harder to remove attributes, though)
And we can minimize burden on code reader by using constants for indexes which
1. would have a name
2. ensure there is no mix up

Reply:
Done.